beacon_node/beacon_chain/src/otb_verification_service.rs

```diff
@@ -0,0 +1,312 @@
+use crate::execution_payload::validate_merge_block;
+use crate::{BeaconChain, BeaconChainError, BeaconChainTypes, BlockError, ExecutionPayloadError};
+use proto_array::InvalidationOperation;
+use slog::{crit, debug, error, info, warn};
+use slot_clock::SlotClock;
+use ssz::{Decode, Encode};
+use ssz_derive::{Decode, Encode};
+use state_processing::per_block_processing::is_merge_transition_complete;
+use std::sync::Arc;
+use store::{DBColumn, Error as StoreError, HotColdDB, KeyValueStore, StoreItem};
+use task_executor::{ShutdownReason, TaskExecutor};
+use tokio::time::sleep;
+use tree_hash::TreeHash;
+use types::{BeaconBlockRef, EthSpec, Hash256, Slot};
+use DBColumn::OptimisticTransitionBlock as OTBColumn;
+
+#[derive(Clone, Debug, Decode, Encode)]
+pub struct OptimisticTransitionBlock {
+    root: Hash256,
+    slot: Slot,
+}
+
+impl OptimisticTransitionBlock {
+    // types::BeaconBlockRef<'_, <T as BeaconChainTypes>::EthSpec>
+    pub fn from_block<T: BeaconChainTypes>(block: BeaconBlockRef<T::EthSpec>) -> Self {
+        Self {
+            root: block.tree_hash_root(),
+            slot: block.slot(),
+        }
+    }
+
+    pub fn root(&self) -> &Hash256 {
+        &self.root
+    }
+
+    pub fn slot(&self) -> &Slot {
+        &self.slot
+    }
+
+    pub fn persist_in_store<T, A>(&self, store: A) -> Result<(), StoreError>
+    where
+        T: BeaconChainTypes,
+        A: AsRef<HotColdDB<T::EthSpec, T::HotStore, T::ColdStore>>,
+    {
+        if store
+            .as_ref()
+            .item_exists::<OptimisticTransitionBlock>(&self.root)?
+        {
+            return Ok(());
+        } else {
+            store.as_ref().put_item(&self.root, self)
+        }
+    }
+
+    pub fn remove_from_store<T, A>(&self, store: A) -> Result<(), StoreError>
+    where
+        T: BeaconChainTypes,
+        A: AsRef<HotColdDB<T::EthSpec, T::HotStore, T::ColdStore>>,
+    {
+        store
+            .as_ref()
+            .hot_db
+            .key_delete(OTBColumn.into(), self.root.as_bytes())
+    }
+
+    fn is_canonical<T: BeaconChainTypes>(
+        &self,
+        chain: &BeaconChain<T>,
+    ) -> Result<bool, BeaconChainError> {
+        Ok(chain
+            // TODO: figure out if this iterator range is inclusive
+            .forwards_iter_block_roots_until(self.slot, self.slot + 1)?
+            .collect::<Result<Vec<_>, _>>()?
+            .into_iter()
+            .map(|(root, _)| root)
+            .collect::<std::collections::HashSet<_>>()
```

Comment:
rather than collecting into a hashset or a vec we could call `next` on the iterator directly and then compare that to `Some(self.root())`. Something like this:

```rust
chain.forwards_iter_block_roots_until(self.slot, self.slot)?
    .next()
    .tranpose()? == Some(self.root())
```

otherwise looks good (the forwards block root iterator _is_ inclusive btw)

Reply:
For reference:

https://github.com/sigp/lighthouse/blob/bb5a6d2cca90931c496b42c4fd11d5c66b21f6cb/beacon_node/beacon_chain/src/beacon_chain.rs#L557-L559

It would be good to document this, if you'd like to add a comment to `forwards_iter_block_roots_until`

Reply:
done! :D